File: cli/src/mcp/tools/get_substrate_context.rs
use serde_json::{json, Value};

use crate::config;
use crate::db;

/// Get I/O recording context for an alert from the InariWatch network.
/// Falls back to local incident memory if the web API is unavailable.
pub async fn execute(args: &Value) -> anyhow::Result<String> {
    let alert_id = args["alert_id"]
        .as_str()
        .ok_or_else(|| anyhow::anyhow!("alert_id is required"))?;

    let conn = db::open()?;
    let alert = db::get_alert_by_id(&conn, alert_id)?
        .ok_or_else(|| anyhow::anyhow!("Alert not found: {}", alert_id))?;

    let cfg = config::load()?;
    let base_url = cfg.global.fix_replay_url.as_deref();

    // Try the web API first if configured
    if let Some(base) = base_url {
        if let Some(fingerprint) = &alert.fingerprint {
            let url = format!(
                "{}/api/recordings/context?fingerprint={}",
                base.trim_end_matches('/'),
                urlencoding::encode(fingerprint)
            );

            let client = reqwest::Client::builder()
                .timeout(std::time::Duration::from_secs(10))
                .build()?;

            if let Ok(resp) = client.get(&url).send().await {
                if resp.status().is_success() {
                    if let Ok(body) = resp.json::<Value>().await {
                        return format_substrate_response(&alert.title, &body);
                    }
                }
            }
        }
    }

    // Fallback: local incident memory for this alert's fingerprint
    if let Some(fingerprint) = &alert.fingerprint {
        let memories = db::get_memories_by_fingerprint(&conn, &alert.project, fingerprint, 3)?;
        if !memories.is_empty() {
            let mut out = format!(
                "Substrate context for alert: {}\n{}\n\n",
                alert.title,
                "=".repeat(50)
            );
            out.push_str("Source: local incident memory (no substrate recording available)\n\n");

            for mem in &memories {
                out.push_str(&format!(
                    "Past fix (confidence: {}%):\n  Root cause: {}\n  Fix: {}\n  Files: {}\n  Worked: {}\n\n",
                    mem.confidence,
                    mem.root_cause,
                    mem.fix_summary,
                    mem.files_fixed.join(", "),
                    if mem.fix_worked { "yes" } else { "no" },
                ));
            }
            return Ok(out);
        }
    }

    // No recording, no memory — return alert context only
    Ok(serde_json::to_string_pretty(&json!({
        "alert_id": alert_id,
        "title": alert.title,
        "body": alert.body,
        "severity": alert.severity,
        "project": alert.project,
        "sources": alert.source_integrations,
        "fingerprint": alert.fingerprint,
        "substrate_recording": null,
        "note": "No substrate recording found. Enable substrate in @inariwatch/capture: INARIWATCH_SUBSTRATE=true"
    }))?)
}

fn format_substrate_response(alert_title: &str, body: &Value) -> anyhow::Result<String> {
    let recording_id = body["recordingId"].as_str().unwrap_or("unknown");
    let command = body["command"].as_str().unwrap_or("unknown");
    let runtime = body["runtime"].as_str().unwrap_or("node");
    let event_count = body["eventCount"].as_u64().unwrap_or(0);
    let duration_ms = body["durationMs"].as_u64().unwrap_or(0);
    let context = body["context"].as_str().unwrap_or("");

    let mut out = format!(
        "Substrate I/O recording for: {}\n{}\n\n",
        alert_title,
        "=".repeat(50)
    );
    out.push_str(&format!(
        "Recording: {}\nCommand: {}\nRuntime: {}\nEvents: {} I/O operations\nDuration: {}ms\n\n",
        recording_id, command, runtime, event_count, duration_ms
    ));

    if !context.is_empty() {
        out.push_str(&format!("Context summary:\n{}\n\n", context));
    }

    if let Some(categories) = body["categories"].as_object() {
        out.push_str("I/O categories:\n");
        for (cat, count) in categories {
            out.push_str(&format!("  • {}: {}\n", cat, count));
        }
        out.push('\n');
    }

    Ok(out)
}

File: web/app/api/mcp/tools/get-substrate-context.ts
import { db, alerts, substrateRecordings } from "@/lib/db";
import { eq } from "drizzle-orm";
import type { McpUser } from "../auth";
import { userCanAccessProject } from "../helpers";

export async function execute(
  args: Record<string, unknown>,
  user: McpUser
): Promise<string> {
  const alertId = args.alert_id as string;
  if (!alertId) return "Error: alert_id is required.";

  // Verify alert belongs to user
  const [alert] = await db
    .select()
    .from(alerts)
    .where(eq(alerts.id, alertId))
    .limit(1);

  if (!alert) return `Error: Alert not found: ${alertId}`;

  const hasAccess = await userCanAccessProject(user.userId, alert.projectId);
  if (!hasAccess) return "Error: Alert does not belong to your projects.";

  // Look for substrate recording linked to this alert
  const [recording] = await db
    .select()
    .from(substrateRecordings)
    .where(eq(substrateRecordings.alertId, alertId))
    .limit(1);

  if (!recording) {
    return JSON.stringify(
      {
        alert_id: alertId,
        title: alert.title,
        body: alert.body?.slice(0, 2000),
        severity: alert.severity,
        ai_reasoning: alert.aiReasoning?.slice(0, 1000),
        sources: alert.sourceIntegrations,
        substrate_recording: null,
        note: "No substrate recording found for this alert. Enable with INARIWATCH_SUBSTRATE=true",
      },
      null,
      2
    );
  }

  let out = `Substrate I/O recording for: ${alert.title}\n${"=".repeat(50)}\n\n`;
  out += `Recording: ${recording.recordingId}\n`;
  out += `Command: ${recording.command ?? "unknown"}\n`;
  out += `Runtime: ${recording.runtime ?? "node"}\n`;
  out += `Events: ${recording.eventCount ?? 0} I/O operations\n`;
  out += `Duration: ${recording.durationMs ?? 0}ms\n\n`;

  if (recording.context) {
    out += `Context:\n${recording.context}\n\n`;
  }

  const categories = recording.categories as Record<string, number> | null;
  if (categories && typeof categories === "object") {
    out += "I/O categories:\n";
    for (const [cat, count] of Object.entries(categories)) {
      out += `  • ${cat}: ${count}\n`;
    }
    out += "\n";
  }

  // Include alert context too
  out += `Alert context:\n`;
  out += `  Severity: ${alert.severity}\n`;
  out += `  Sources: ${(alert.sourceIntegrations ?? []).join(", ")}\n`;
  if (alert.aiReasoning) {
    out += `  AI reasoning: ${alert.aiReasoning.slice(0, 500)}\n`;
  }

  return out;
}
